### ISSUE TITLE
Type conversion of OPTION_SELECT_ALL constant value in build-options module for the ModelSelect component results in values not being selected

### ISSUE THREAD

**Issue by tn3rb** [eventespresso/event-espresso-core#546]
Discovered a minor gotcha with the `build-options` module for the `ModelSelect` component when used for selecting query arguments (like filtering event attendees by datetime)...

we use the following constant:

```js
export const OPTION_SELECT_ALL = 'ALL';
```

to indicate that any/all values in the list are acceptable.

The problem occurs when the value for the list is an integer (like a datetime ID) and you are performing logic on that value that expects it to be an integer,
like `parseInt()` or comparisons like `value > 0`, etc. As well, if you are saving the value in a block attribute as an integer then nothing gets recorded when the incoming value is `'ALL'`.

When React then feeds a value of `null` or `undefined` back to the select component, the `'ALL'` option does not get selected since it does not match and the component stays on whatever the previous value was, ie: displays the last selected element instead of displaying the ALL options.

This can be worked around when saving the attribute value in your block by simply converting any non-absolute integer values to `'ALL'` and then storing the value as a string instead of an integer. This also requires you to convert and non-`'ALL'` values back to integer when passing values to the `ModelSelect`.

So basic type juggling shenanigans.

We may want to consider implementing some other method for dealing with this that can avoid having to juggle the data types around.

One option may be to convert `OPTION_SELECT_ALL` into a property instead of a constant and then allow each child component to set its own value for it.
So if your selector uses integer values then you could set it to `0` or `-1` and if your selector used string values then you could use `'ALL'` or whatever appropriate value.

Another option may be to use two constants, like:

```js
export const OPTION_SELECT_ALL_STRING = 'ALL';
export const OPTION_SELECT_ALL_NUMBER = 0;
```

And and use the appropriate constant value in the options list either by detecting the incoming element values programmatically or by allowing the child component to explicitly set it via another property.

The benefit to doing this within the bounds of the `build-options` module and the `ModelSelect` component is that it cuts down on constantly having to account for this gotcha every time the component is used.

**To Reproduce**
Implement one of the model select components like `DatetimeSelect` in a GB block and save the `datetimeId` attribute value as an integer.
In your `setAttributes` callback, perform any kind of logic that expects the incoming value to be an integer, ex:

```js
setDatetimeId = ( datetimeId ) => {
	this.props.setAttributes( { datetimeId: parseInt( datetimeId.value, 10 ) } );
};
```

then attempt to choose the "Select All Datetimes" option from the list. 
Selecting any other valid option that uses an integer value will display correctly but will not change when selecting the "All" option

**Expected behavior**
choosing a "Select All" option should result in that option being displayed in the list.

**Comment by nerrad** [comment]
Hmm ya, this is a tricky one.  Part of the dilemma here is that in most cases ALL is really just an alias for "no filtering" (or an equivalent to "none").   So I wonder if we even NEED a "ALL" option, and instead parent components implementing the selectors could just change the default text for no selection to be labelled (All Datetimes).  I think this would be the better solution.

**Comment by tn3rb** [comment]
that won't work because there needs to be an actual option to select to indicate no selection. 
Placeholder text only works when the input has no value because no option has even been selected. 
Once that happens and the input has a value, the placeholder text disappears and there is no way to "unselect" options and return the input to having no value.

**Comment by nerrad** [comment]
With react-select, if the `defaultValue` of zero (or no value is) passed to it is treated as no selection and thus the placeholder text shows (if you're not seeing that behaviour, there might be a setup issue in your branch).  If there's nothing selected, then the current setup of our select hocs will not include the filtering information.  

I'm not sure how "unselect" options is needed here because if it defaults to having no value then what is there to unselect.

So I still think it'd work.

**Comment by nerrad** [comment]
So in discussing things in slack, I realized that a feature we configured  for react-select (an `x` ui element for clearing the select) was not implemented fully.  In `ModelSelect` I need to listen for the `onChange` and capture the `action` callback to handle the `clear` event.

**Comment by nerrad** [comment]
fixed in #549
